opentype-shaping-gurmukhi.md

```diff
@@ -1029,12 +1031,63 @@ so that glyphs of the same ordering category remain in the same
 relative position with respect to each other.
 
 
+#### 2.10: Flag sequences for possible feature applications ####
+
+With the inital reordering complete, those glyphs in the syllable that
+may have GSUB or GPOS features applied in stages 3, 5, and 6 should be
+flagged for each potential feature. 
+
+This flagging is preliminary; the set of potential features varies
+between different scripts and which features are supported varies
+between fonts. It is also possible that the application of
+one feature on a glyph sequence will perform a substitution that makes
+a later feature no longer applicable to the updated sequence.
+
+Consequently, the flagging must be completed before shaping procedes
+to the stages during which features are applied.
+
+Some shaping features, such as `locl`, can potentially apply to any
+glyphs. Therefore it is not necessary to maintain a separate flag for
+these features in the bitmask (or other data structure) used to track
+the flags -- although shaping engines may do so if desired.
+
+The sequences to flag are summarized in the list below; a full
+description of each feature's function and interpretation is provided
+in GSUB and GPOS application stages that follow.
+
+  - `nukt` should match "_Consonant_,Nukta" sequences
+  - `akhn` should match "Ka,Halant,Ssa" and "Ja,Halant,Nya"
+  - `rphf` should match initial "Ra,Halant" sequences but _not_ match
+            initial "Ra,Halant,ZWJ" sequences
+  - `blwf` should match "Halant,Ra", "Halant,Ha", and "Halant,Va" in
+            post-base positions and "Ra,Halant", "Ha,Halant", and
+            "Va,Halant" in non-initial pre-base positions
+  - `half` should match "_Consonant_,Halant" in pre-base position but
+           _not_ match "Ra,Halant" sequences flagged for `rphf` and
+           _not_ match "_Consonant_,Halant,ZWNJ,_Consonant_" sequences
+  - `pstf` should match initial "Halant,Ya" in post-base position
+  - `vatu` should match "_Consonant_,Halant,Ra",
+           "_Consonant_,Halant,Ha", and "_Consonant_,Halant,Va"
+  - `cjct` should match "_Consonant_,Halant,_Consonant_" but _not_
+            match "_Consonant_,Halant,ZWJ,_Consonant_" or
+            "_Consonant_,Halant,ZWNJ,_Consonant_"
+  - `init` should match word-initial letters
+  - `haln` should match word-final "_Consonant_,Halant" sequences
+
+> Note: Word-initial consonants and independent vowels will be flagged
+> for possible `init` feature application here. Any matras that are
+> reordered to a word-initial position during the final-reordering
+> stage will be flagged for possible `init` feature application
+> separately, after final reordering.
```

Comment:
I don't think pre-base matras are ever reordered to a word-initial position during final reordering. Rather, pre-base matras that were already moved to a word-initial position during initial reordering can potentially be shuffled forward closer to the base consonant during final reordering.

> Word-initial consonants and independent vowels will be flagged for possible `init` feature application here.

Are there examples of this in Indic? I don't recall HarfBuzz or Prince doing this.

Reply:
> I don't think pre-base matras are ever reordered to a word-initial position during final reordering. Rather, pre-base matras that were already moved to a word-initial position during initial reordering can potentially be shuffled forward closer to the base consonant during final reordering.

Well this may be a wording thing. Not necessarily that the left-matra would move, but that Reph would be reordered and leave the left-matra in word-initial position as a result. This is a reference to 4.5; does the (main branch) 4.5 not match what Price does?

> > Word-initial consonants and independent vowels will be flagged for possible `init` feature application here.
> 
> Are there examples of this in Indic? I don't recall HarfBuzz or Prince doing this.

Similar question here; 4.5 currently just addresses left-matras (solo or decomposed ones, implicitly). Should it also address all other letters?

To be honest, we are playing in rather uncharted water with `init` anyway. The Microsoft docs technically only call it mandatory for Bengali, although that seems to be one of those decisions the linguist camps might call "overly prescriptive".

Reply:
> Not necessarily that the left-matra would move, but that Reph would be reordered and leave the left-matra in word-initial position as a result.

Ah, of course :facepalm:! I'd failed to consider left matras moving to a word-initial position by virtue of other characters moving out of the way.

> > Are there examples of this in Indic? I don't recall HarfBuzz or Prince doing this.
>
> Similar question here; 4.5 currently just addresses left-matras (solo or decomposed ones, implicitly). Should it also address all other letters?

I'm not sure, I'll defer to your expertise on this. As an aside, this sentence:

> Word-initial consonants and independent vowels will be flagged for possible `init` feature application here.

leads me to believe that the flagging of word-initial consonants and independent vowels should be done in 2.10. Regardless of what `init` should apply to, can't all the flagging just be done in 4.5? Or have I failed to consider some edge cases here?

Reply:
> I'm not sure, I'll defer to your expertise on this. As an aside, this sentence:
> 
> > Word-initial consonants and independent vowels will be flagged for possible `init` feature application here.
> 
> leads me to believe that the flagging of word-initial consonants and independent vowels should be done in 2.10. Regardless of what `init` should apply to, can't all the flagging just be done in 4.5? Or have I failed to consider some edge cases here?

I guess HarfBuzz only actually uses 4.5 to add a flag for `init` on left matras, and otherwise dumps `init` and `haln` into the "basic features" that it applies at the very end. But it has been a little while since I really read through all of HarfBuzz's layers to feel sure I know where the "basic features" are executed. I had thought that, from the bug discussion earlier, Allsorts was doing it earlier, but that may be my misreading the wording.

Conceptually, I guess it doesn't matter until then (4.5).

Perhaps I am just worried that we need to mention the word-initial and word-final conditional ones _some_where, because that is stuff that happens at a higher-level context than "syllable" -- the context where, obviously, everything else is being decided. So glossing over it could mislead someone.

Reply:
> I guess HarfBuzz only actually uses 4.5 to add a flag for init on left matras, and otherwise dumps `init` and `haln` into the "basic features" that it applies at the very end. But it has been a little while since I really read through all of HarfBuzz's layers to feel sure I know where the "basic features" are executed. I had thought that, from the bug discussion earlier, Allsorts was doing it earlier, but that may be my misreading the wording.

It's less likely your misreading, and more my inability to express myself precisely.

As far as I'm aware, both HarfBuzz and Allsorts follow the sequence of steps 2-5 as laid out in this spec. So the ordering of these steps is definitely not in question, and apologies that I lead you to believe otherwise!

> Note: Word-initial consonants and independent vowels will be flagged
> for possible `init` feature application here. Any matras that are
> reordered to a word-initial position during the final-reordering
> stage will be flagged for possible `init` feature application
> separately, after final reordering.

What I was questioning was my understanding of the note, which I read as:

1. Flag word-initial consonants and independent vowels for possible `init` in **2.10**.
2. Flag word-initial matras for possible `init` in **4.5**.

I'd never heard of `init` being applied to word-initial consonants or independent vowels (HarfBuzz definitely doesn't). But if indeed it can/should, my question is: regardless of what `init` should apply to, can the flagging of `init` just be done in 4.5, as opposed to being split across two different steps (2.10 and 4.5)?

Reply:
> What I was questioning was my understanding of the note, which I read as:

Ah; gotcha. Back on the same page now.

> 
>     1. Flag word-initial consonants and independent vowels for possible `init` in **2.10**.
> 
>     2. Flag word-initial matras for possible `init` in **4.5**.
> 
> 
> I'd never heard of `init` being applied to word-initial consonants or independent vowels (HarfBuzz definitely doesn't). But if indeed it can/should, my question is: regardless of what `init` should apply to, can the flagging of `init` just be done in 4.5, as opposed to being split across two different steps (2.10 and 4.5)?

OK; this is an odd one.... For the life of me, I really thought that `init` was defined in `bng2` for all characters, even though the Microsoft guidance page now clearly says it is only for left-side matras.

Turns out, that might have been an idea I got from the old version of the page. John Hudson spearheaded a rewrite of the feature definition in 2016 [(http://tiro.com/OTWG/20160203-Joining_Feature_Proposal_1.2.pdf)](http://tiro.com/OTWG/20160203-Joining_Feature_Proposal_1.2.pdf) to better align it with its primary intended usage in Arabic & joining scripts.

But, up through at least 2017 ([Wayback link](https://web.archive.org/web/20171203161902/http://www.microsoft.com:80/typography/otfntdev/bengalot/features.htm)) the page mentioned it applying to consonants. By sheer coincidence, I had read through that material when working on Bengali during my MATD year, so it likely stuck in my head then....

I only see `init` substitutions for left matras in Noto and Murty, so I think I will record this oddity in an issue, but make the appropriate change to left-matra–onlyness in the doc.

Reply:
Change made, and issue open at #104 ... if any later consensus emerges about clear guidance regard `init` scope it can be revisited.